### PR TITLE
Dual certificate support (eg ECDSA with RSA fallback)

### PR DESCRIPTION
**Pull request by polarathene**
# Description

Continuation of my original PR: https://github.com/docker-mailserver/docker-mailserver/pull/1629

> **TL;DR:** This feature enables support for multiple certificates, eg for serving a modern ECDSA cert with an RSA cert as fallback.

See the original PR and/or commit messages for added details. Additionally (from original PR), **there's also fixes for bugs related to modifying internal cert configs**.

**NOTE:** *This does not enable support for ECDSA certificates*. Only the ability to serve a primary certificate with a fallback (in future there will be EdDSA that differs from ECDSA which would in future replace RSA as a fallback). ECDSA support is dependent upon a separate PR updating the TLS ciphersuites.

## Potential blockers:

- [x] There [may still be a concern to address with Dovecot](https://github.com/docker-mailserver/docker-mailserver/pull/1629#issuecomment-703400592)
- [x] Tests (need guidance or assistance)
- [x] Docs

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or ENVIRONMENT.md or the Wiki)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aendeavor** [comment]
Tests obviously need some adjustments as they also check some thing related to TLS levels and cipher suites. But that shouldn't be a big deal - just changing the defaults in the tests to use the new, appropriate and updated suites and levels.

File: target/scripts/start-mailserver.sh [hunk 1]
```diff
@@ -1147,6 +1147,51 @@ function _setup_ssl
 {
   _notify 'task' 'Setting up SSL'
 
+  # Primary certificate to serve for TLS
+  # NOTE: The `sed` substituion delimiter uses `~` instead of `/` due to file paths as values
+  function _set_certificate()
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1152]: Everything looks good to me when it comes to Bash style. I guess there is one or more commit coming - therefore changing `function _set_certificate()` to `function _set_certificate` can be implemented in one of the following commits. I know, nip-picky here - but these scripts need great attention :)

**Comment by polarathene** [comment]
For tests I need advice on existing tests that need to be updated or added. I'm not familiar with the testing infrastructure, and it's directory is full of many directories and files. One of my commits updates a test for letsencrypt:

> `@test "checking ssl: letsencrypt configuration is correct"`
> `#test domain has certificate files`
> `#test hostname has certificate files`

For both of those it removes two separate tests (that individually configure key or cert files), and updates to the new config setting (`smtpd_tls_chain_files`, postfix only) which takes a sequence of files containing the private key and cert.

Perhaps I need to write something here that does similar for the fallback certificate? (postfix uses the same new config setting but includes both primary and fallback, dovecot has two new config settings specifically for fallback cert)

---

> Tests obviously need some adjustments as they also check some thing related to TLS levels and cipher suites.

I assume this is about ECDSA cipher suites? That's fair, but as this doesn't enable those cipher suites (my other PR does), I think tests for this just need to pass the current certificate related tests and add some new tests that somehow test the fallback support is working? (basically, if the client doesn't accept ECDSA, they should still be able to accept RSA certificate with cipher suites using RSA digital signature)

> just changing the defaults in the tests to use the new, appropriate and updated suites and levels.

I think you got this mixed up with my other PR? This one is just about dual certificate support (and fixes / improvements on `start-mailserver.sh` config issues).

Perhaps getting the related PR merged first is a good idea? Probably can't properly test this feature without it.

**Comment by polarathene** [comment]
Regarding the removal of the `-combined.pem` file for Dovecot when using `self-signed`.

I've found that it was [added to `generate-ssl-certificate` in 2015](https://github.com/docker-mailserver/docker-mailserver/commit/ea830f5ec608c8de0320df16919cdb31fa18bec2) when Courier was used (required a fullchain bundle file) instead of Dovecot. When Dovecot replaced it, `-combined.pem` was still used instead of `-cert.pem` like Postfix was using. **No additional action required**, just documenting the change better.

---

[`generate-ssl-certificate`](https://github.com/docker-mailserver/docker-mailserver/blob/f168e3bd716e077941fde001810fad8c60854f70/target/bin/generate-ssl-certificate#L22-L32) appears to also have comments and roughly the same commands as described in the [Postfix docs for setting up a private CA](http://www.postfix.org/TLS_README.html#private-ca).

**`cacert.pem` does not appear to be used for anything else beyond postfix with `self-signed`** (Dovecot also can support specifying a CA, but when it replaced Courier.. this was missed). I don't think it's required that a self-signed CA be added to sign a single self-signed certificate. This is probably legacy as well and can be dropped.

By doing so, we could remove the `generate-ssl-certificate` file/command which requires user interaction for adding details, but as `self-signed` mode is strictly meant for non-production usage, we shouldn't require those details to be accurate. `manual` mode can be used by someone who insists otherwise.

There's a test TODO task related to `self-signed` / `generate-ssl-certificate`:

https://github.com/docker-mailserver/docker-mailserver/blob/f168e3bd716e077941fde001810fad8c60854f70/test/tests.bats#L1443-L1447

**Should we remove that along with the `setup.sh` command `config ssl`?** The cert generation code could be migrated to `start-mailserver.sh` and live there instead for `self-signed` mode, implicitly setting up the certificate.

---

These changes could be deferred to another PR if that's more suitable.

**Comment by aendeavor** [comment]
> I think you got this mixed up with my other PR?

Indeed.

> Perhaps getting the related PR merged first is a good idea? 

Indeed as well. I will mark this as a draft again :)

---

> **Should we remove that along with the setup.sh command config ssl?**

I think this would be a good idea, since we are going to publish this with version `8.1.0`, this is fine. A bump in minor version bump can justify this change.

> These changes could be deferred to another PR if that's more suitable.

If it's more comfortable for you.

---

About the tests: Is there anything you have a question about? The infrastructure is best-described here: https://github.com/docker-mailserver/docker-mailserver/pull/1643#issuecomment-722631834

**Comment by aendeavor** [comment]
@polarathene Out of draft state since #1802 was merged :)

**Comment by aendeavor** [comment]
What do you think @polarathene, how fast can you get this to be ready for merging? We are assessing whether this should be in `v9.0.0` or not.

**Comment by polarathene** [comment]
> how fast can you get this to be ready for merging?

I just need to look into why tests are failing:

<details>
<summary>Failing test snippets</summary>

```
not ok 80 checking postfix: remove privacy details of the sender
# (in test file test/mail_privacy.bats, line 43)
#   `docker exec mail_privacy /bin/sh -c "openssl s_client -quiet -starttls smtp -connect 0.0.0.0:587 < /tmp/docker-mailserver-test/email-templates/send-privacy-email.txt"' failed
# 140464342774912:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1544:SSL alert number 40
```

```sh
@test "checking postfix: remove privacy details of the sender" {
  docker exec mail_privacy /bin/sh -c "openssl s_client -quiet -starttls smtp -connect 0.0.0.0:587 < /tmp/docker-mailserver-test/email-templates/send-privacy-email.txt"
```

---

```
 not ok 139 checking saslauthd: ldap smtp authentication
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
#  in test file test/mail_with_ldap.bats, line 210)
#   `assert_success' failed
# 
# -- command failed --
# status : 1
# output : 140615162684544:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1544:SSL alert number 40
# --
# 
```

```sh
@test "checking saslauthd: ldap smtp authentication" {
  run docker exec mail_with_ldap /bin/sh -c "nc -w 5 0.0.0.0 25 < /tmp/docker-mailserver-test/auth/sasl-ldap-smtp-auth.txt | grep 'Authentication successful'"
  assert_success
  run docker exec mail_with_ldap /bin/sh -c "openssl s_client -quiet -connect 0.0.0.0:465 < /tmp/docker-mailserver-test/auth/sasl-ldap-smtp-auth.txt | grep 'Authentication successful'"
  assert_success
```

---

```
 not ok 267 checking system: /var/log/mail/mail.log is error free
# (from function `assert_failure' in file test/test_helper/bats-assert/src/assert.bash, line 140,
#  in test file test/tests.bats, line 516)
#   `assert_failure' failed
# 
# -- command succeeded, but it was expected to fail --
# output : Feb 18 10:13:27 mail postfix/submission/smtpd[3651]: warning: TLS library problem: error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher:../ssl/statem/statem_srvr.c:2259:
# --
# 
```

```sh
@test "checking system: /var/log/mail/mail.log is error free" {
  run docker exec mail grep 'non-null host address bits in' /var/log/mail/mail.log
  assert_failure
  run docker exec mail grep 'mail system configuration error' /var/log/mail/mail.log
  assert_failure
  run docker exec mail grep ': error:' /var/log/mail/mail.log
  assert_failure
```

---

```
not ok 312 checking setup.sh: setup.sh dkim help
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
#  in test file test/tests.bats, line 1132)
#   `assert_success' failed
# 
# -- command failed --
# status : 1
# output (65 lines):
#   setup.sh Bootstrapping Script
#   
#   Usage: ./setup.sh [-i IMAGE_NAME] [-c CONTAINER_NAME] <subcommand> <subcommand> [args]
#   
#   OPTIONS:
#   
#     -i IMAGE_NAME     The name of the docker-mailserver image
#                       The default value is
#                       'docker.io/mailserver/docker-mailserver:latest'
#   
#     -c CONTAINER_NAME The name of the running container.
#   
#     -p PATH           Config folder path (default: /home/runner/work/docker-mailserver/docker-mailserver/config)
#   
#     -h                Show this help dialogue
#   
#     -z                Allow container access to the bind mount content
#                       that is shared among multiple containers
#                       on a SELinux-enabled host.
#   
#     -Z                Allow container access to the bind mount content
#                       that is private and unshared with other containers
#                       on a SELinux-enabled host.
#   
#   SUBCOMMANDS:
#   
#     email:
#   
#       ./setup.sh email add <email> [<password>]
#       ./setup.sh email update <email> [<password>]
#       ./setup.sh email del [OPTIONS] <email>
#       ./setup.sh email restrict <add|del|list> <send|receive> [<email>]
#       ./setup.sh email list
#   
#     alias:
#   
#       ./setup.sh alias add <email> <recipient>
#       ./setup.sh alias del <email> <recipient>
#       ./setup.sh alias list
#   
#     quota:
#   
#       ./setup.sh quota set <email> [<quota>]
#       ./setup.sh quota del <email>
#   
#     config:
#   
#       ./setup.sh config dkim [keysize <size>] [domain '<domain1.tld>[,<domain2.tld>...]']
#       ./setup.sh config ssl <fqdn>
#   
#     relay:
#   
#       ./setup.sh relay add-domain <domain> <host> [<port>]
#       ./setup.sh relay add-auth <domain> <username> [<password>]
#       ./setup.sh relay exclude-domain <domain>
#   
#     debug:
#   
#       ./setup.sh debug fetchmail
#       ./setup.sh debug fail2ban [<unban> <ip-address>]
#       ./setup.sh debug show-mail-logs
#       ./setup.sh debug inspect
#       ./setup.sh debug login <commands>
#   
#     help: Show this help dialogue
# --
# 
```

```sh
@test "checking setup.sh: setup.sh dkim help" {
  run ./setup.sh -c mail dkim help
  assert_success
```

---

```
not ok 325 checking postfix submission: only A grade TLS ciphers are used
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
#  in test file test/tests.bats, line 1260)
#   `assert_success' failed
# 
# -- command failed --
# status : 1
# output : 
# --
# 
```

```sh
# postfix submission TLS
@test "checking postfix submission: only A grade TLS ciphers are used" {
  run docker run --rm -i --link mail:postfix \
    --entrypoint sh instrumentisto/nmap -c \
      'nmap --script ssl-enum-ciphers -p 587 postfix | grep "least strength: A"'
  assert_success
}
```

---

```
ok 326 checking postfix submission: nmap produces no warnings on TLS ciphers verifying
not ok 327 checking postfix smtps: only A grade TLS ciphers are used
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
#  in test file test/tests.bats, line 1276)
#   `assert_success' failed
# 
# -- command failed --
# status : 1
# output : 
# --
# 
```

```sh
@test "checking postfix submission: nmap produces no warnings on TLS ciphers verifying" {
  run docker run --rm -i --link mail:postfix \
    --entrypoint sh instrumentisto/nmap -c \
      'nmap --script ssl-enum-ciphers -p 587 postfix | grep "warnings" | wc -l'
  assert_success
  assert_output 0
}

# postfix smtps SSL
@test "checking postfix smtps: only A grade TLS ciphers are used" {
  run docker run --rm -i --link mail:postfix \
    --entrypoint sh instrumentisto/nmap -c \
      'nmap --script ssl-enum-ciphers -p 465 postfix | grep "least strength: A"'
  assert_success
}
```

</details>

Other than that, adding another cipher list test shouldn't be an issue. Just need to handle the Dovecot concerned mentioned above and add env var docs. Should be good before monday?

**I'm not sure why these tests are failing however..** I'll look into them but any assistance would be helpful. eg for the last two tests that test  `'nmap --script ssl-enum-ciphers -p <port> postfix | grep "least strength: A"'` (where `<port>` is 587 or 465), I'm unable to reproduce this error out of the CI at present.

The `tests.bats` file is fairly large, and the initialization in `setup()` is a fair bit, I haven't reproduced that locally yet, perhaps I'll run the test file with all other test cases stripped.

**Comment by polarathene** [comment]
Slimmed `tests.bats` down to just the failing test cases, and checked `docker logs mail` while that ran to see if anything useful was there:

> mail postfix/smtpd[2831]: warning: error opening chain file: /etc/ssl/certs/ssl-fullkeychain-snakeoil.pem: No such file or directory

I guess that means my PR is at fault under a certain configuration :sweat_smile:

**Comment by polarathene** [comment]
This was due to my config change of:

```
smtpd_tls_cert_file=/etc/ssl/certs/ssl-cert-snakeoil.pem
smtpd_tls_key_file=/etc/ssl/private/ssl-cert-snakeoil.key
```
with non-existent file:

```
smtpd_tls_chain_files = /etc/ssl/certs/ssl-fullkeychain-snakeoil.pem
```

I thought that by the name `snakeoil` these were fake files that didn't actually exist, and if they did that they weren't sourced officially from somewhere. [Turns out they are](https://askubuntu.com/questions/396120/what-is-the-purpose-of-the-ssl-cert-snakeoil-key), but they're still not intended to be used in production.

Technically this failure should be valid, and TLS shouldn't be enabled as the `start-mailserver.sh` indicates should be the case if no `SSL_TYPE` is given.

Should probably change the [handling from `ssl=required` to `ssl=yes`](https://github.com/docker-mailserver/docker-mailserver/blob/9efa94ce6f5ff51e34cdf19e3497a69031e330b4/target/scripts/start-mailserver.sh#L1330), to instead change to [`ssl=no` for dovecot](https://doc.dovecot.org/configuration_manual/dovecot_ssl_configuration/), and similarly disable TLS for postfix too?

That or I can use these snakeoil files as they were handled before? (and perhaps outputting a clear warning about insecure setup, since the current log for this case doesn't imply a bad SSL setup).

**Comment by aendeavor** [comment]
> Should probably change the [handling from `ssl=required` to `ssl=yes`](https://github.com/docker-mailserver/docker-mailserver/blob/9efa94ce6f5ff51e34cdf19e3497a69031e330b4/target/scripts/start-mailserver.sh#L1330), to instead change to [`ssl=no` for dovecot](https://doc.dovecot.org/configuration_manual/dovecot_ssl_configuration/), and similarly disable TLS for postfix too?
> 
> That or I can use these snakeoil files as they were handled before? (and perhaps outputting a clear warning about insecure setup, since the current log for this case doesn't imply a bad SSL setup).

TBH, I don't know what's better. I would leave it up to you to decide. This change can be breaking too, since it will be included in the next major release `v9.0.0`.

**Comment by polarathene** [comment]
Not sure what's going on here but it seems ClamAV in `setup()` is failing. I can see that memory usage climbs up to 2GB+ (from ~300MB), then it drops and this failure output is given:

```sh
./test/bats/bin/bats test/tests.bats
 ✗ 
   (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 37,
    from function `repeat_in_container_until_success_or_timeout' in file test/test_helper/common.bash, line 70,
    from function `setup_file' in test file test/tests.bats, line 53)
     `repeat_in_container_until_success_or_timeout 60 mail test -e /var/run/clamav/clamd.ctl' failed
   26d3295fc834e4e959f5f4959439240de6958f36f0804f1c81583ab98b7969b1
   [ TASKLOG ]  mail.my-domain.com is up and running
   Timed out on command: docker exec mail test -e /var/run/clamav/clamd.ctl
   mail
   bats warning: Executed 1 instead of expected 6 tests

6 tests, 1 failure, 5 not run
```

It looks like a timeout is one of the expected outcomes, but this is failing the tests before they run for me, as this is run in `setup()`, is that intentional? I'm not sure of it's relevance for the test cases in `tests.bats`.

**Comment by aendeavor** [comment]
Not sure what happened here, but at least on `8.0.1` ClamAV works. I will test on `:edge` now.

**Comment by polarathene** [comment]
> EDIT: Seems to work on `:edge` too. I just tested with `supervisorctl status clamav` and checked whether `/Var/run/clamav/clamd.ctl` is there.

This is not building locally right? (git clone of repo, init with submodules, `docker build -t mailserver-testing:ci .`, running the `make` command to prep testing data and running `tests.bats` against the newly built image)

I assume that there must be something different with the CI environment. Test is updated, sorting out docs now, might have a few other minor changes, then I'll add some release note snippets for you and we'll be good for review and merge :)

**Comment by aendeavor** [comment]
> This is not building locally right? (git clone of repo, init with submodules, `docker build -t mailserver-testing:ci .`, running the `make` command to prep testing data and running `tests.bats` against the newly built image)

You're probably missing build arguments. Have a look at `make build`, which essentially wraps `docker build ...`.

> I assume that there must be something different with the CI environment.

Not really. CI executes `make build` in the same way.

> you and we'll be good for review and merge :)

Very nice.

**Comment by polarathene** [comment]
> You're probably missing build arguments. Have a look at `make build`, which essentially wraps `docker build ...`.

Nope, like I said I ran the make command as the CI seems to, I don't see anything missing there. I'll investigate another time.

> Not really. CI executes `make build` in the same way.

Seems [other users had the issue in the past](https://github.com/docker-mailserver/docker-mailserver/issues/6), I was testing with a 1GB VPS that had plenty of swap but perhaps I need to spin up a larger VPS instance to confirm if it was memory pressure related.

---

Related to this PR, [`SSL_TYPE=custom`](https://github.com/docker-mailserver/docker-mailserver/pull/90) came before [`SSL_TYPE=manual`](https://github.com/docker-mailserver/docker-mailserver/pull/296) in the same year (2016). Prior there was only `letsencrypt` and `self-signed` support.

I'm going to look into refactoring / consolidating the `SSL_TYPE` handling in a future PR. I think we have some unnecessary redundancy shoved in complicating things for no reason and either no one was bothered enough or not comfortable touching this area of the script so I'll be happy to tackle that :)

Like wise, the [`Wiki - Configure SSL`](https://github.com/docker-mailserver/docker-mailserver/wiki/Configure-SSL) docs aren't being touched with this PR (not that they can be in a PR atm). I'll be tackling the wiki migration into this repo for static doc gen and hosting next, after that I'll revise the SSL docs.

**Comment by aendeavor** [comment]
SGTM

**Comment by polarathene** [comment]
Should be good to go now :+1: 

Reviewing may be easier via commits if the full PR diff is difficult to work through. I personally like Gitkraken as a GUI where you can easily navigate through the commits or select multiple commits at once to view an aggregated diff with different viewing layouts.

---

## Release Notes

- **[fix] TLS:** For some `docker-compose` setups when restarting the `docker-mailserver` container, internal config state may have been persisted despite making changes that should reconfigure TLS (eg changing `SSL_TYPE` or replacing the certificate file).
- **[feat] TLS:** Dual (aka hybrid) certificate support! (eg `ECDSA` certificate with an `RSA` fallback for broader compatibility)
  - This feature is presently only for `SSL_TYPE=manual`, all you need to do is provide your fallback certificate to the `SSL_ALT_CERT_PATH` and `SSL_ALT_KEY_PATH` ENV vars, just like your primary certificate would be setup for `manual` mode.

File: target/scripts/start-mailserver.sh [hunk 2]
```diff
@@ -1138,6 +1138,62 @@ function _setup_ssl
...
+      DOVECOT_CERT="${CERT_CHAIN}"
+    fi
+
+    # Postfix configuration
+    # NOTE: `smtpd_tls_chain_files` expects private key defined before public cert chain
+    # May be a single PEM file or a sequence of files, so long as the order is key->leaf->chain
+    sed -i 's~^smtpd_tls_chain_files =.*~smtpd_tls_chain_files = '"${POSTFIX_KEY_WITH_FULLCHAIN}~" /etc/postfix/main.cf
+
+    # Dovecot configuration
+    sed -i 's~^ssl_key = <.*~ssl_key = <'"${DOVECOT_KEY}~" /etc/dovecot/conf.d/10-ssl.conf
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1166]: Noticed that this `sed` should actually be:

``` BASH
sed -i "s|^ssl_key = <.*|ssl_key = <${DOVECOT_KEY}|" /etc/dovecot/conf.d/10-ssl.conf
```

Which goes for the `sed` below too.

File: target/scripts/start-mailserver.sh [hunk 3]
```diff
@@ -1255,85 +1317,93 @@ function _setup_ssl
...
-        sed -i -e 's~ssl_cert = </etc/dovecot/ssl/dovecot\.pem~ssl_cert = </etc/postfix/ssl/'"${HOSTNAME}"'-combined\.pem~g' /etc/dovecot/conf.d/10-ssl.conf
-        sed -i -e 's~ssl_key = </etc/dovecot/ssl/dovecot\.key~ssl_key = </etc/postfix/ssl/'"${HOSTNAME}"'-key\.pem~g' /etc/dovecot/conf.d/10-ssl.conf
+        cp /tmp/docker-mailserver/ssl/demoCA/cacert.pem /etc/postfix/ssl
+        # Have Postfix trust the self-signed CA (which is not installed within the OS trust store)
+        sed -i -r 's~^#?smtpd_tls_CAfile =.*~smtpd_tls_CAfile = /etc/postfix/ssl/cacert.pem~' /etc/postfix/main.cf
+        sed -i -r 's~^#?smtp_tls_CAfile =.*~smtp_tls_CAfile = /etc/postfix/ssl/cacert.pem~' /etc/postfix/main.cf
+        # Part of the original `self-signed` support, unclear why this symlink was required?
+        # May have been to support the now removed `Courier` (Dovecot replaced it):
+        # https://github.com/docker-mailserver/docker-mailserver/commit/1fb3aeede8ac9707cc9ea11d603e3a7b33b5f8d5
+        local PRIVATE_CA='/etc/ssl/certs/cacert-'"${HOSTNAME}"'.pem'
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1387]: Again, using

``` BASH
local PRIVATE_CA="/etc/ssl/certs/cacert-${HOSTNAME}.pem"
```

is the right way :D

File: target/scripts/start-mailserver.sh [hunk 4]
```diff
@@ -1138,6 +1138,62 @@ function _setup_ssl
...
+    sed -i 's~^smtpd_tls_chain_files =.*~& '"${PRIVATE_KEY_ALT} ${CERT_CHAIN_ALT}~" /etc/postfix/main.cf
+
+    # Dovecot configuration
+    # Conditionally checks for `#`, in the event that internal container state is accidentally persisted,
+    # can be caused by: `docker-compose up` run again after a `ctrl+c`, without running `docker-compose down`
+    sed -i 's~^#\?ssl_alt_key = <.*~ssl_alt_key = <'"${PRIVATE_KEY_ALT}"'~' /etc/dovecot/conf.d/10-ssl.conf
+    sed -i 's~^#\?ssl_alt_cert = <.*~ssl_alt_cert = <'"${CERT_CHAIN_ALT}"'~' /etc/dovecot/conf.d/10-ssl.conf
+  }
+
   function _apply_tls_level()
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1197]: Please remove the `()` from the function definition :)

File: target/scripts/start-mailserver.sh [hunk 5]
```diff
@@ -1138,6 +1138,62 @@ function _setup_ssl
...
+    # NOTE: This operation doesn't replace the line, it appends to the end of the line.
+    # Thus this method should only be used when this line has explicitly been replaced earlier in the script.
+    # Otherwise without `docker-compose down` first, a `docker-compose up` may
+    # persist previous container state and cause a failure in postfix configuration.
+    sed -i 's~^smtpd_tls_chain_files =.*~& '"${PRIVATE_KEY_ALT} ${CERT_CHAIN_ALT}~" /etc/postfix/main.cf
+
+    # Dovecot configuration
+    # Conditionally checks for `#`, in the event that internal container state is accidentally persisted,
+    # can be caused by: `docker-compose up` run again after a `ctrl+c`, without running `docker-compose down`
+    sed -i 's~^#\?ssl_alt_key = <.*~ssl_alt_key = <'"${PRIVATE_KEY_ALT}"'~' /etc/dovecot/conf.d/10-ssl.conf
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1193]: As above already mentioned, misuse of `''"` and `''`. Please correct :)

File: target/scripts/start-mailserver.sh [hunk 6]
```diff
@@ -1138,6 +1138,62 @@ function _setup_ssl
...
+    cp "${COPY_CERT_FROM_PATH}" "${CERT_CHAIN_ALT}"
+    chmod 600 "${PRIVATE_KEY_ALT}"
+    chmod 600 "${CERT_CHAIN_ALT}"
+
+    # Postfix configuration
+    # NOTE: This operation doesn't replace the line, it appends to the end of the line.
+    # Thus this method should only be used when this line has explicitly been replaced earlier in the script.
+    # Otherwise without `docker-compose down` first, a `docker-compose up` may
+    # persist previous container state and cause a failure in postfix configuration.
+    sed -i 's~^smtpd_tls_chain_files =.*~& '"${PRIVATE_KEY_ALT} ${CERT_CHAIN_ALT}~" /etc/postfix/main.cf
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1188]: As above already mentioned, misuse of `""` and `''` in `sed`. Please correct :)

File: target/scripts/start-mailserver.sh [hunk 7]
```diff
@@ -1234,13 +1295,14 @@ function _setup_ssl
...
         _notify 'inf' "Adding ${LETSENCRYPT_DOMAIN} SSL certificate to the postfix and dovecot configuration"
 
-        # Postfix configuration
-        sed -i -r 's~smtpd_tls_cert_file=/etc/ssl/certs/ssl-cert-snakeoil.pem~smtpd_tls_cert_file=/etc/letsencrypt/live/'"${LETSENCRYPT_DOMAIN}"'/fullchain.pem~g' /etc/postfix/main.cf
-        sed -i -r 's~smtpd_tls_key_file=/etc/ssl/private/ssl-cert-snakeoil.key~smtpd_tls_key_file=/etc/letsencrypt/live/'"${LETSENCRYPT_DOMAIN}"'/'"${LETSENCRYPT_KEY}"'\.pem~g' /etc/postfix/main.cf
+        # LetsEncrypt `fullchain.pem` and `privkey.pem` contents are detailed here from CertBot:
+        # https://certbot.eff.org/docs/using.html#where-are-my-certificates
+        # `key.pem` was added for `simp_le` support (2016): https://github.com/docker-mailserver/docker-mailserver/pull/288
+        # `key.pem` is also a filename used by the `_extract_certs_from_acme` method (implemented for Traefik v2 only)
+        local PRIVATE_KEY='/etc/letsencrypt/live/'"${LETSENCRYPT_DOMAIN}"'/'"${LETSENCRYPT_KEY}"'.pem'
```
Comment (aendeavor) [inline target/scripts/start-mailserver.sh:1302]: Please adjust the `""` and `''` here too.

**Comment by aendeavor** [comment]
Other than a few styling and apostrophes issues, this LGTM

File: target/scripts/start-mailserver.sh (hunk 2 above)
Comment (polarathene) [reply target/scripts/start-mailserver.sh:1166]: I noticed you changed the `sed` delimiter from `~` to `|` is this a style convention for the project? They were previously `/` but that was changed by me due to the same character being used in filepath values. `_apply_tls_level` still kept the `/` delimiter for example.

Comment (aendeavor) [reply target/scripts/start-mailserver.sh:1166]: We are currently using `|` or `+` for `sed`, but honest to goodness, it's not that important actually:) I would be nice, but the `""` thing is more important. I figured, when you're at it, you can go ahead an change the delimiters too.

Comment (polarathene) [reply target/scripts/start-mailserver.sh]: I updated the `sed` substitution delimiters in the file (not just related to this PR), replacing any usage of `/` or `~`. After CI tests pass, should be good for merge.

Comment (polarathene) [reply target/scripts/start-mailserver.sh]: Oops, missed the merge window :sweat_smile: I'll add it as an additional PR then.

**Pull request merged**